source/common/upstream/upstream_impl.cc

```diff
@@ -799,6 +799,43 @@ ClusterInfoImpl::ClusterInfoImpl(
 
     lb_type_ = LoadBalancerType::ClusterProvided;
     break;
+  case envoy::config::cluster::v3::Cluster::LOAD_BALANCING_POLICY_CONFIG: {
+    if (config.has_lb_subset_config()) {
+      throw EnvoyException(
+          fmt::format("cluster: LB policy {} cannot be combined with lb_subset_config",
+                      envoy::config::cluster::v3::Cluster::LbPolicy_Name(config.lb_policy())));
+    }
+
+    if (config.has_common_lb_config()) {
+      throw EnvoyException(
+          fmt::format("cluster: LB policy {} cannot be combined with common_lb_config",
+                      envoy::config::cluster::v3::Cluster::LbPolicy_Name(config.lb_policy())));
+    }
+
+    if (!config.has_load_balancing_policy()) {
+      throw EnvoyException(
+          fmt::format("cluster: LB policy {} requires load_balancing_policy to be set",
+                      envoy::config::cluster::v3::Cluster::LbPolicy_Name(config.lb_policy())));
+    }
+
+    const auto& lb_policy =
+        std::find_if(config.load_balancing_policy().policies().begin(),
+                     config.load_balancing_policy().policies().end(),
+                     [](envoy::config::cluster::v3::LoadBalancingPolicy_Policy policy) {
+                       return Registry::FactoryRegistry<TypedLoadBalancerFactory>::getFactory(
+                                  policy.name()) != nullptr;
+                     });
+
+    if (lb_policy == config.load_balancing_policy().policies().end()) {
+      throw EnvoyException(fmt::format(
+          "Didn't find a registered load balancer factory implementation for cluster: '{}'",
+          name_));
+    }
```

Comment:
This might not needed when we use getAndCheckFactoryByName appropriately.

Reply:
I don't think the cluster manager should be responsible for picking the LB policy out of `load_balancing_policies`. IMO it's better if the cluster info does this instead - then it knows _which_ policy was picked, and doesn't have to carry around config for any policies that weren't picked.

Reply:
Why we want to do a) check whether the lb policy factory registered , and b) create lb policy from factory in two different places? That doesn't make sense either.

You can pick LB policy in ClusterManagerImpl and store a reference to the lb policy factory in a member in ClusterInfoImpl, and just use that in cluster manager.

Reply:
I think I'm missing an important detail somewhere... I'm under the impression that we can't store the factory we've picked from the `ClusterManagerImpl`, since `info()` returns a `const` pointer. By the time the cluster manager picks the LB policy it's too late to write what we picked into the `ClusterInfoImpl`. That's why I was hoping to pick the policy in the `ClusterInfoImpl` constructor instead.

This is mostly a result of wanting to be able to write a unit test that verifies that the set of policies { 1. policy with no registered factory, 2. policy with a registered factory } will result in the known policy being picked. It seems to me this requires that the `ClusterInfoImpl` to know which policy was picked, and hence to do the picking itself. I suppose this problem disappears if we don't bother with that bit of the test but it seems to me like a reasonably important behaviour to ensure is correct here.

Arguably a) is not so much about checking whether the factory is registered as much as it is about actually picking the policy - which happens to guarantee as a postcondition that the chosen factory is registered by definition. So I'm not how problematic it is to have policy picking in the `ClusterInfoImpl` constructor and the actual LB instantiation in the cluster manager instead. What do you think?

Reply:
> I think I'm missing an important detail somewhere... I'm under the impression that we can't store the factory we've picked from the `ClusterManagerImpl`, since `info()` returns a `const` pointer. By the time the cluster manager picks the LB policy it's too late to write what we picked into the `ClusterInfoImpl`. That's why I was hoping to pick the policy in the `ClusterInfoImpl` constructor instead.

I meant you can implement lb_policy_factory() in ClusterInfoImpl to return the reference to the factory. My point here is to reduce manual check config and throw EnvoyException to make the error message more consistent and easy to evolve extension API in general, that's why I pointed to just use TypedExtensionConfig in proto part as well.